File: backend/services/tax_service.go
package services

import (
	"context"
	"mastutik-api/dto"
	"mastutik-api/models"
	"mastutik-api/repositories"
)

type TaxService interface {
	CreateTax(ctx context.Context, req dto.CreateTaxRequest) (*models.Tax, error)
	GetAllTaxes(ctx context.Context) ([]models.Tax, error)
	GetActiveTaxes(ctx context.Context) ([]models.Tax, error)
	UpdateTax(ctx context.Context, id uint, req dto.UpdateTaxRequest) (*models.Tax, error)
	DeleteTax(ctx context.Context, id uint) error
}

type taxService struct {
	repo repositories.TaxRepository
}

func NewTaxService(repo repositories.TaxRepository) TaxService {
	return &taxService{repo: repo}
}

func (s *taxService) CreateTax(ctx context.Context, req dto.CreateTaxRequest) (*models.Tax, error) {
	tax := &models.Tax{
		Name:         req.Name,
		Percentage:   req.Percentage,
		ActiveStatus: req.ActiveStatus,
	}
	err := s.repo.Create(ctx, tax)
	return tax, err
}

func (s *taxService) GetAllTaxes(ctx context.Context) ([]models.Tax, error) {
	return s.repo.FindAll(ctx)
}

func (s *taxService) GetActiveTaxes(ctx context.Context) ([]models.Tax, error) {
	return s.repo.FindActive(ctx)
}

func (s *taxService) UpdateTax(ctx context.Context, id uint, req dto.UpdateTaxRequest) (*models.Tax, error) {
	tax, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}

	if req.Name != nil {
		tax.Name = *req.Name
	}
	if req.Percentage != nil {
		tax.Percentage = *req.Percentage
	}
	if req.ActiveStatus != nil {
		tax.ActiveStatus = *req.ActiveStatus
	}

	err = s.repo.Update(ctx, tax)
	return tax, err
}

func (s *taxService) DeleteTax(ctx context.Context, id uint) error {
	return s.repo.Delete(ctx, id)
}

File: frontend/src/services/tax.service.ts
import { apiClient } from '../lib/api-client';
import { Tax } from '../types';

export const taxService = {
  // Public
  getActiveTaxes: async (): Promise<Tax[]> => {
    const res = await apiClient.get('/taxes');
    return res.data;
  },

  // Admin
  getAllTaxes: async (): Promise<Tax[]> => {
    const res = await apiClient.get('/admin/taxes');
    return res.data;
  },

  createTax: async (data: { name: string; percentage: number; active_status: boolean }): Promise<Tax> => {
    const res = await apiClient.post('/admin/taxes', data);
    return res.data;
  },

  updateTax: async (id: number, data: { name?: string; percentage?: number; active_status?: boolean }): Promise<Tax> => {
    const res = await apiClient.put(`/admin/taxes/${id}`, data);
    return res.data;
  },

  deleteTax: async (id: number): Promise<void> => {
    await apiClient.delete(`/admin/taxes/${id}`);
  },
};
